wgpu-core/src/command/pass.rs

```diff
@@ -0,0 +1,337 @@
+        .binder
+        .assign_group(index as usize, bind_group, &state.temp_offsets);
+    if !entries.is_empty() && pipeline_layout.is_some() {
+        let pipeline_layout = pipeline_layout.as_ref().unwrap().raw();
+        for (i, e) in entries.iter().enumerate() {
+            if let Some(group) = e.group.as_ref() {
+                let raw_bg = group.try_raw(state.snatch_guard)?;
+                unsafe {
+                    state.raw_encoder.set_bind_group(
+                        pipeline_layout,
+                        index + i as u32,
+                        Some(raw_bg),
+                        &e.dynamic_offsets,
+                    );
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// After a pipeline has been changed, resources must be rebound
+pub(crate) fn rebind_resources<E, F: FnOnce()>(
+    state: &mut BaseState,
+    pipeline_layout: &Arc<binding_model::PipelineLayout>,
+    late_sized_buffer_groups: &[LateSizedBufferGroup],
+    f: F,
+) -> Result<(), E>
+where
+    E: From<DestroyedResourceError>,
+{
+    if state.binder.pipeline_layout.is_none()
+        || !state
+            .binder
+            .pipeline_layout
+            .as_ref()
+            .unwrap()
+            .is_equal(pipeline_layout)
+    {
+        let (start_index, entries) = state
+            .binder
+            .change_pipeline_layout(pipeline_layout, late_sized_buffer_groups);
+        if !entries.is_empty() {
+            for (i, e) in entries.iter().enumerate() {
+                if let Some(group) = e.group.as_ref() {
+                    let raw_bg = group.try_raw(state.snatch_guard)?;
+                    unsafe {
+                        state.raw_encoder.set_bind_group(
+                            pipeline_layout.raw(),
+                            start_index as u32 + i as u32,
+                            Some(raw_bg),
+                            &e.dynamic_offsets,
+                        );
+                    }
+                }
+            }
+        }
+
+        f();
+
+        let non_overlapping =
+            super::bind::compute_nonoverlapping_ranges(&pipeline_layout.push_constant_ranges);
+
+        // Clear push constant ranges
+        for range in non_overlapping {
+            let offset = range.range.start;
+            let size_bytes = range.range.end - offset;
+            super::push_constant_clear(offset, size_bytes, |clear_offset, clear_data| unsafe {
+                state.raw_encoder.set_push_constants(
+                    pipeline_layout.raw(),
+                    range.stages,
+                    clear_offset,
+                    clear_data,
+                );
+            });
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn set_push_constant<E, F: FnOnce(&[u32])>(
+    state: &mut BaseState,
+    push_constant_data: &[u32],
+    stages: wgt::ShaderStages,
+    offset: u32,
+    size_bytes: u32,
+    values_offset: u32,
+    f: F,
+) -> Result<(), Option<E>>
```

Comment:
```suggestion
    values_offset: Option<u32>,
    f: F,
) -> Result<(), E>
```

Same here where `E: From<PushConstantUploadError> + From<InvalidValuesOffset> + From<MissingPipeline>`.